roachtest: cdc/poller/rangefeed=false failed

[(roachtest).cdc/poller/rangefeed=false failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2721381&tab=buildLog) on [master@6601d827b814d4e85a1081b03bf2562d8ac2a4ab](https://github.com/cockroachdb/cockroach/commits/6601d827b814d4e85a1081b03bf2562d8ac2a4ab):

```
		  | (1) signal: killed
		  | Error types: (1) *exec.ExitError
		Wraps: (5) context canceled
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *main.withCommandDetails (4) *secondary.withSecondaryError (5) *errors.errorString

	cluster.go:2688,cdc.go:207,cdc.go:611,test_runner.go:767: monitor failure: unexpected node event: 3: dead
		(1) attached stack trace
		  -- stack trace:
		  | main.(*monitor).WaitE
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2676
		  | main.(*monitor).Wait
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2684
		  | main.cdcBasicTest
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:207
		  | main.registerCDC.func3
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cdc.go:611
		  | main.(*testRunner).runTest.func2
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:767
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (2) monitor failure
		Wraps: (3) unexpected node event: 3: dead
		Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *errors.errorString

	cluster.go:1667,context.go:140,cluster.go:1656,test_runner.go:848: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2721381-1614322964-28-n4cpu16 --oneshot --ignore-empty-nodes: exit status 1 4: skipped
		3: dead
		1: 5139
		2: 4856
		Error: UNCLASSIFIED_PROBLEM: 3: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1147
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:271
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1852
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 3: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/cdc/poller/rangefeed=false](https://teamcity.cockroachdb.com/viewLog.html?buildId=2721381&tab=artifacts#/cdc/poller/rangefeed=false)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Acdc%2Fpoller%2Frangefeed%3Dfalse.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

I see a panic:
```
I210226 08:11:07.749799 1 util/log/file_sync_buffer.go:238 â‹® [config]   file created at: 2021/02/26 08:11:07
I210226 08:11:07.749826 1 util/log/file_sync_buffer.go:238 â‹® [config]   running on machine: â€¹teamcity-2721381-1614322964-28-n4cpu16-0003â€º
I210226 08:11:07.749838 1 util/log/file_sync_buffer.go:238 â‹® [config]   binary: CockroachDB CCL v21.1.0-alpha.3-879-g6601d827b8-dirty (x86_64-unknown-linux-gnu, built 2021/02/26 06:46:17, go1.15.6)
I210226 08:11:07.749846 1 util/log/file_sync_buffer.go:238 â‹® [config]   arguments: â€¹[./cockroach start --insecure --store=path=/mnt/data1/cockroach --log-dir=/home/ubuntu/logs --cache=25% --max-sql-memory=25% --port=26257 --http-port=26258 --locality=cloud=gce,region=us-central1,zone=us-central1-b --join=104.154.162.0:26257 --advertise-host=10.128.0.169]â€º
I210226 08:11:07.749869 1 util/log/file_sync_buffer.go:238 â‹® [config]   line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg utf8=âœ“
I210226 08:11:07.749432 1 util/log/flags.go:195  [-] 1  stderr capture started
panic: resolved timestamp regression, was 1614327531.833072282,1, recomputed as 1614327531.833072282,0 [recovered]
	panic: resolved timestamp regression, was 1614327531.833072282,1, recomputed as 1614327531.833072282,0
```

--- a/pkg/kv/kvserver/rangefeed/resolved_timestamp.go
+++ b/pkg/kv/kvserver/rangefeed/resolved_timestamp.go
@@ -217,13 +217,27 @@ func (rts *resolvedTimestamp) recompute() bool {
 	if !rts.IsInit() {
 		return false
 	}
-	newTS := rts.closedTS
-	if txn := rts.intentQ.Oldest(); txn != nil {
-		txnTS := txn.timestamp.FloorPrev()
-		if txnTS.Less(newTS) {
-			newTS = txnTS
-		}
+	if rts.closedTS.Less(rts.resolvedTS) {
+		panic(fmt.Sprintf("closed timestamp below resolved timestamp: %s < %s",
+			rts.closedTS, rts.resolvedTS))
 	}
+	newTS := rts.closedTS
+
+	// Take into account the intents that haven't been yet resolved - their
+	// timestamps cannot be resolved yet.
+	if txn := rts.intentQ.Oldest(); txn != nil {
+		if txn.timestamp.LessEq(rts.resolvedTS) {
+			panic(fmt.Sprintf("unresolved txn equal to or below resolved timestamp: %s <= %s",
+				txn.timestamp, rts.resolvedTS))
+		}
+		// txn.timestamp cannot be resolved, so the resolved timestamp must be Prev.
+		txnTS := txn.timestamp.Prev()
+		newTS.Backward(txnTS)
+	}
+	// Truncate the logical part. It might have come from a Prev call above, and
+	// it's dangerous to start pushing things above Logical=MaxInt32.
+	newTS.Logical = 0
+
 	if newTS.Less(rts.resolvedTS) {
 		panic(fmt.Sprintf("resolved timestamp regression, was %s, recomputed as %s",
 			rts.resolvedTS, newTS))

--- a/pkg/kv/kvserver/rangefeed/resolved_timestamp_test.go
+++ b/pkg/kv/kvserver/rangefeed/resolved_timestamp_test.go
@@ -16,6 +16,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
 	"github.com/stretchr/testify/require"
 )
@@ -562,4 +563,44 @@ func TestResolvedTimestampTxnAborted(t *testing.T) {
 	fwd = rts.ConsumeLogicalOp(abortTxnOp(txn1))
 	require.True(t, fwd)
 	require.Equal(t, hlc.Timestamp{WallTime: 25}, rts.Get())
+}
+
+// Test that things go well when the closed timestamp has non-zero logical part.
+func TestClosedTimestampLogicalPart(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	rts := makeResolvedTimestamp()
+	rts.Init()
+
+	// Set a new closed timestamp. Resolved timestamp advances.
+	fwd := rts.ForwardClosedTS(hlc.Timestamp{WallTime: 10, Logical: 2})
+	require.True(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 10, Logical: 0}, rts.Get())
+
+	// Add an intent for a new transaction.
+	txn1 := uuid.MakeV4()
+	fwd = rts.ConsumeLogicalOp(writeIntentOp(txn1, hlc.Timestamp{WallTime: 10, Logical: 4}))
+	require.False(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 10, Logical: 0}, rts.Get())
+
+	// Set a new closed timestamp. Resolved timestamp doesn't advance, since it
+	// could only theoretically advance up to 10.4, and it doesn't do logical
+	// parts.
+	fwd = rts.ForwardClosedTS(hlc.Timestamp{WallTime: 11, Logical: 6})
+	require.False(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 10, Logical: 0}, rts.Get())
+
+	// Abort txn1. Resolved timestamp advances.
+	fwd = rts.ConsumeLogicalOp(abortTxnOp(txn1))
+	require.True(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 11, Logical: 0}, rts.Get())
+
+	// Create an intent one tick above the closed ts. Resolved timestamp doesn't
+	// advance. This tests the case where the closed timestamp has a logical part,
+	// and an intent is in the next wall tick; this used to cause an issue because
+	// of the rounding logic.
+	txn2 := uuid.MakeV4()
+	fwd = rts.ConsumeLogicalOp(writeIntentOp(txn2, hlc.Timestamp{WallTime: 12, Logical: 7}))
+	require.False(t, fwd)
+	require.Equal(t, hlc.Timestamp{WallTime: 11, Logical: 0}, rts.Get())
 }

